### PR TITLE
Forbid Var name collisions in a kustomization stack

### LINKED ISSUE #634: Vars with same name at different levels have no defined behavior
A Var Name collision in either one Kustomization file or at different levels in a Kustomization overlay/base stack doesn't doesn't trigger any kind of reporting or warning, and has no documented behavior.

We should either enforce a scoping rule (e.g. vars in the overlay override vars they shadow in the base), or
disallow collisions.

In the interest of simplicity, suggest the latter - so a var has the same value all the way up the stack (no surprises).

### PR DESCRIPTION
**Pull request by monopole**
Test added in first commit to repro bug, fix is in second commit.

Fixes #634 
Groundwork for #506 and #605

### CODE REVIEW

File: pkg/target/kusttarget_test.go [hunk 1]
```diff
@@ -336,3 +336,171 @@ bases:
...
+`)
+	vars, err := makeKustTarget(t, ldr).getAllVars()
+	if err != nil {
+		t.Fatalf("Err: %v", err)
+	}
+	if len(vars) != 4 {
+		t.Fatalf("unexpected size %d", len(vars))
+	}
+	v2 := someVars[2]
+	v2.Name = "HOT"
```
Comment (Liujingfang1) [inline pkg/target/kusttarget_test.go]: What is the purpose of this test?  Overlay o1 and base both have env `HOT`. Do we pick up the overlay one or base one? Is this a conflict?

File: pkg/target/kusttarget.go [hunk 2]
```diff
@@ -124,22 +124,17 @@ func (kt *KustTarget) MakeCustomizedResMap() (resmap.ResMap, error) {
...
 		}
 	}
-
 	// Given that names have changed (prefixs/suffixes added),
 	// fix all the back references to those names.
-	err = transformers.NewNameReferenceTransformer(
-		kt.tConfig.NameReference).Transform(m)
-	if err != nil {
-		return nil, err
+	if kt.tConfig.NameReference != nil {
```
[the remark this answers is not in the document]
Comment (monopole) [reply pkg/target/kusttarget.go]: 1) Postel’s Law (robustness) and 2) so tests aren't obligated to do it.

If we want to enforce that this not be nil, we should do so when reading the configs.  Perhaps we do, i didn't look.

**Pull request merged**